mkosi/__init__.py

```diff
@@ -504,47 +515,55 @@ def cache_params(state: MkosiState, root: Path) -> list[PathString]:
     return flatten(("--bind", state.cache, root / p) for p in state.installer.cache_path())
 
 
+def mount_build_overlay(state: MkosiState) -> ContextManager[Path]:
+    return mount_overlay(state.root, state.build_overlay, state.workdir, state.root)
+
+
 def run_prepare_script(state: MkosiState, cached: bool) -> None:
     if state.config.prepare_script is None:
         return
     if cached:
         return
 
-    verb = "build" if state.do_run_build_script else "final"
-
-    with complete_step("Running prepare script…"):
-        bwrap: list[PathString] = [
-            "--bind", state.config.build_sources, "/root/src",
-            "--bind", state.config.prepare_script, "/root/prepare",
-            *cache_params(state, Path("/")),
-            "--chdir", "/root/src",
-        ]
-
-        run_workspace_command(state, ["/root/prepare", verb], network=True, bwrap_params=bwrap,
-                              env=dict(SRCDIR="/root/src"))
+    bwrap: list[PathString] = [
+        "--bind", state.config.build_sources, "/root/src",
+        "--bind", state.config.prepare_script, "/root/prepare",
+        *cache_params(state, Path("/")),
+        "--chdir", "/root/src",
+    ]
 
+    def clean() -> None:
         srcdir = state.root / "root/src"
         if srcdir.exists():
             srcdir.rmdir()
 
         state.root.joinpath("root/prepare").unlink()
 
+    with complete_step("Running prepare script…"):
+        run_workspace_command(state, ["/root/prepare", "final"], network=True, bwrap_params=bwrap,
+                              env=dict(SRCDIR="/root/src"))
+        clean()
+
+    if state.config.build_script:
+        with complete_step("Running prepare script in build overlay…"), mount_build_overlay(state):
+            run_workspace_command(state, ["/root/prepare", "build"], network=True, bwrap_params=bwrap,
+                                env=dict(SRCDIR="/root/src"))
```

Comment:
The indentation of `env` is off, but preferably we could 
```suggestion
            run_workspace_command(
                state, ["/root/prepare", "build"], network=True, bwrap_params=bwrap, env=dict(SRCDIR="/root/src")
            )
```
do some black-like indentation here, because otherwise it's easy to miss the `mount_build_overlay` there (as happened to me).